app/src/main/java/com/ncnf/database/DatabaseService.java

```diff
@@ -1,125 +1,245 @@
+        CompletableFuture<List<T>> futureResponse = new CompletableFuture<>();
+
+        this.db.collection(collectionPath).orderBy(field).startAt(value).endAt(value + "\uf8ff").get().addOnCompleteListener(task -> {
+            if(task.isSuccessful()){
+                List<T> result = new ArrayList<>();
+                for(DocumentSnapshot document : task.getResult().getDocuments()){
+                    result.add((T) registry.get(collectionType).toObject(document.getId(), document.getData()));
+                }
+
+                futureResponse.complete(result);
+            } else {
+                futureResponse.completeExceptionally(task.getException());
+            }
+        });
 
-    public CompletableFuture<DatabaseResponse> removeArrayField(String path, String arrayField, Object value){
-        return this.updateField(path, arrayField, FieldValue.arrayRemove(value));
+        return futureResponse;
     }
 
-    public <U> CompletableFuture<DatabaseResponse> whereEqualTo(String path, String field, List<U> values) {
-        if(values.size() == 0) throw new IllegalArgumentException("You must pass at least one value.");
+    @Override
+    public <T, R> CompletableFuture<List<R>> whereArrayContains(String collectionPath, String field, T value, Class<R> type) {
+        CompletableFuture<List<R>> futureResponse = new CompletableFuture<>();
 
-        CompletableFuture<DatabaseResponse> futureResponse = new CompletableFuture<>();
-        CollectionReference ref = this.db.collection(path);
-        Query query = ref.whereEqualTo(field, values.get(0));
-        for (int i = 1; i < values.size(); ++i) {
-            query = query.whereEqualTo(field, values.get(i));
-        }
-        query.get().addOnCompleteListener(task -> onTaskComplete(task, futureResponse));
+        this.db.collection(collectionPath).whereArrayContains(field, value).get().addOnCompleteListener(task -> {
+            if(task.isSuccessful()){
+                List<R> result = new ArrayList<>();
+                for(DocumentSnapshot document : task.getResult().getDocuments()){
+                    result.add((R) registry.get(type).toObject(document.getId(), document.getData()));
+                }
+
+                futureResponse.complete(result);
+            } else {
+                futureResponse.completeExceptionally(task.getException());
+            }
+        });
 
         return futureResponse;
     }
 
-    protected void onTaskComplete(Task task, CompletableFuture<DatabaseResponse> futureResponse){
-        try {
-            if(task.getResult() instanceof DocumentSnapshot){
-                futureResponse.complete(new DatabaseResponse(task.isSuccessful(), ((DocumentSnapshot) task.getResult()).getData(), task.getException()));
+    @Override
+    public <T, R> CompletableFuture<List<R>> whereEqualTo(String collectionPath, String field, T value, Class<R> type) {
+        CompletableFuture<List<R>> futureResponse = new CompletableFuture<>();
```

Comment:
At this point I'm starting to wonder whether you could have had a generic function to perform the query, which would have been a high order function, taking the query function as argument

Reply:
I thought about that but for a good abstraction we should keep it this way since the DatabaseService is the only one which has the Instance of FirebaseFirestore...